### ISSUE TITLE
Feature - Show output model logits

### ISSUE THREAD

**Issue by freQuensy23-coder** [ollama/ollama#2415]
Feature request:
How can i get logits (probabilites of each next token), during generation, just like I can do it in Open AI API (logprobs)? This feature will be helpfull for apps, that use logprobs to measure model avareness and confidence.

**Comment by neychevr** [comment]
Hello! Seems like a really good feature for more complex usage of Ollama.
Is this feature already WIP, or contribution would be fine?

**Comment by mateon1** [comment]
I would also like to have this, I'm interested in having both echo + logprobs, so I can get information about the prompt too, instead of just the completion. Right now I'm using very small models with pytorch to compute logits directly, but that's really slow.

**Comment by magic-YuanTian** [comment]
Any updates?

**Comment by briancleland** [comment]
> Any updates?

https://github.com/ollama/ollama/pull/1640#issuecomment-2043381653

**Comment by SharmaM-dev** [comment]
Any updates?

**Comment by drdsgvo** [comment]
Are there any updates on this very important issue? To not implement logits is not a valid solution. Anyone (including me) who needs logits will move from Ollama to a different solution! Please be aware of that.

**Comment by The-Inscrutable-X** [comment]
support, would be very nice

**Comment by Morinator** [comment]
I'm surprised this is not one of the first things implemented 🤔

**Comment by haukelicht** [comment]
Hi there,

any progress on integrating this feature request?

**Comment by szocsbarni** [comment]
Hi, is there a timeline available for integration?

**Comment by mommi84** [comment]
> Hi, is there a timeline available for integration?

We are going to get AGI before this, 100%!

**Comment by SharmaM-dev** [comment]
[laugh]         Mridul Sharma reacted to your message:
________________________________

**Comment by martinkozle** [comment]
We are going to get GTA 6 before this.

**Comment by latent-variable** [comment]
man, I really need this for to implement an CoT-Decoding pipeline. I guess I'll go back to playing Sparking Zero.

**Comment by josiahbryan** [comment]
I've given up hope and switched back to llama.cpp for production inference.
Using it with ramalama which can pull from the ollama model library.

Really disappointed that the maintainers here show such disregard for such
a huge community request.

 Makes me want to make sure I don't use the project in any way. If an
obvious thing like this is being totally ignored by the maintainers, then
it shows they don't really care much about what the community is asking
for.

**Comment by briancleland** [comment]
@jmorganca @bmizerany Has the team given up on implementing this feature?

**Comment by NumberChiffre** [comment]
Pls make this happen lol, as a painful user on Mac :(

**Comment by josiahbryan** [comment]
@jmorganca @bmizerany you guys broadcast your partnership with Hugging Face - great! What about this though? This seems like less than 1/10th the effort - why are you ignoring everyone asking for input here? Why don't you at least provide a timeline?

**Comment by athmanar** [comment]
insane that this not given as an option? maybe better to switch to pure huggingface models

**Comment by codelion** [comment]
> man, I really need this for to implement an CoT-Decoding pipeline for the open-webui. I guess I'll go back to playing Sparking Zero.

Cot decoding and entropy decoding are available in optillm - https://github.com/codelion/optillm

**Comment by Cy-Fi** [comment]
Ollama will stop being an option for us if crucial features like this are not beeing implemented...

**Comment by magic-YuanTian** [comment]
For such a simple but important feature, the team demonstrates unexpected arrogance and ignorance. I think this is a red flag for us to give up using Ollama as LLM backend since they cannot go far.

**Comment by codelion** [comment]
I have implemented it in PyTorch if anyone is looking for it they can use the following colab - https://colab.research.google.com/drive/1zPv47_tog2_KOFJY-WJxwPYR6mgoxKlK?usp=sharing 

Here is the discussion on optillm where it was brought up as well - https://github.com/codelion/optillm/discussions/82

**Comment by codelion** [comment]
Logprobs are now directly supported in our local inference server - https://github.com/codelion/optillm?tab=readme-ov-file#local-inference-server we use the OpenAI compatible API so you can get them using the same code.

**Comment by jooray** [comment]
This would be very useful for enforcing the structure of output (output_cls with langchain, that currently works with llama.cpp and hugging face).

It can reject tokens that would break the output structure. Very useful for tool calling as well.

**Comment by drdsgvo** [comment]
> Logprobs are now directly supported in our local inference server - https://github.com/codelion/optillm?tab=readme-ov-file#local-inference-server we use the OpenAI compatible API so you can get them using the same code.

Great to hear that we have a cool alternative to ollama as those guys are not doing what needs to be done!

**Comment by jooray** [comment]
I would suggest you being more kind. Ollama is an open source project, they are not working for you. Feel free to offer a bounty to implement this, or create a pull request.

I would really like to see this implemented, but that does not mean I have to be mean to authors of a software I get for free. And being an ass in comments (which many of you are here) is not very motivating for developers either.

**Comment by ParthSareen** [comment]
Hey everyone! Sorry for the delay and no updates here - going to be picking this up soon and hopefully getting it in early Jan. There's been a ton of changes on the API even more so coming on the inference engine layer so just need to be a bit careful as it would be an API addition but it is something we want to support!

**Comment by ParthSareen** [comment]
Hey folks would like to get your thoughts:
Would you care if it was logits vs logprobs? Would you prefer one over the other? If so why? 

Working on designing the API and getting the functionality right along with that :) Appreciate your patience!

**Comment by josiahbryan** [comment]
Personally would prefer log probs, just because all my tooling is setup for
that and used to thinking in logorobs haha

**Comment by martinkozle** [comment]
But if you want to calculate the probability of the LLM generating "yes" or "no" for example you would either have to use constrained generation with logprobs, where only those 2 tokens will be non 0. Or you can directly use logits and do the constraining yourself.

**Comment by mommi84** [comment]
> Would you care if it was logits vs logprobs? Would you prefer one over the other? If so why?

Definitely logprobs so that it doesn't deviate from the OpenAI standards ([see examples here](https://cookbook.openai.com/examples/using_logprobs)) and the following can be supported:

```python
for token in response.choices[0].logprobs.content:
    for top_logprob in token.top_logprobs:
        print((top_logprob.token, np.exp(top_logprob.logprob)))
```

**Comment by ParthSareen** [comment]
Thanks @josiahbryan, @martinkozle, @mommi84. I also think it makes more sense to have logprobs for now and then happy to re-evaluate when doing the new engine. Have some fun plans for sampling :)

**Comment by Elimane0800** [comment]
> Hey folks would like to get your thoughts:
> 
> Would you care if it was logits vs logprobs? Would you prefer one over the other? If so why? 
> 
> 
> 
> Working on designing the API and getting the functionality right along with that :) Appreciate your patience!

Personally I prefer logits because we can obtain ourselves a kind of logprobs if we have logits. Plus when trying tasks such as distillation or uncertainty observation, logits are more interesting to have. So please add logits 🙏, we can do the logprobs calculations by ourselves

**Comment by Elimane0800** [comment]
To be more precise : Logprobs can be derived directly from logits through a softmax followed by a logarithmic transformation. This kind of operation is neither the most difficult nor the most time consuming. It's easy to get logprobs from logits but not that easy to get logits from logprobs except if you add a scale constant to get an approximation of the logits value.

**Comment by Elimane0800** [comment]
Sorry for writing 3 successive message. This one should probably be another feature request. I'll definitely open one. Is it possible to add dropout layer to the models (like Monte Carlo dropout for example) ? Plus access the neural network layer information of models.

**Comment by BenjaminMarechalEVITECH** [comment]
I disagree with @Elimane0800 , to get logprobs from logits, one need to compute logsoftmax on logits of the complete vocabulary, which is an expensive operation and is already done in the model for the inference of the next token.
Moreover, it is often only necessary to get the N most probable tokens (with N of the order of a few dozen). In this case, logprobs (or probs) are relevant, logits are not.

**Comment by Elimane0800** [comment]
> I disagree with @Elimane0800 , to get logprobs from logits, one need to compute logsoftmax on logits of the complete vocabulary, which is an expensive operation and is already done in the model for the inference of the next token.
> 
> Moreover, it is often only necessary to get the N most probable tokens (with N of the order of a few dozen). In this case, logprobs (or probs) are relevant, logits are not.

For having already tried to calculate logprobs from logits with cpu i can say that it is definitely not the type of task that will run during 24h. So if the only matter is a question of time and cost of the operation I can tell that it is nor the most time consuming neither the most expensive operation.

**Comment by BenjaminMarechalEVITECH** [comment]
There are two use cases:
- we want to know the probabilities of the N most probable tokens
- we want to know the probabilities or the logits of all the tokens in the vocabulary.

I think the first use case is the most requested today. To answer it effectively, the API has no choice but to give the logprobs (or probs) of the N most probable tokens. Indeed, if the API only provides the logits, then it must provide them for the entire vocabulary if we want to deduce the logprobs. With vocabulary sizes sometimes approaching 100k, this overloads the API's JSON response enormously.

**Comment by ParthSareen** [comment]
Hey folks, thanks for your patience! Going to be doing something like top k log probs for the API. Not saying no to logits, but given the comments, API design, and system constraints it just makes sense to do this first.

There's also some adjacent work going on right now as well in which there's some API refactoring/design to be done. Upon it's completion this will be one of the first things that go out :)

**Comment by OriginalGoku** [comment]
any timeframe for this update?

**Comment by codelion** [comment]
> any timeframe for this update?

It may take a while to get it here since they rely on llama.cpp underneath. You can try an alternative 
 - https://github.com/ollama/ollama/issues/2415#issuecomment-2453442288

**Comment by ClaudiuCreanga** [comment]
> It may take a while to get it here since they rely on llama.cpp underneath. You can try an alternative like optillm - https://github.com/codelion/optillm
> 
> * [Provide logits or logprobs in the API #2415 (comment)](https://github.com/ollama/ollama/issues/2415#issuecomment-2453442288)

Seems like this one is an open ai tooling, while we're interested in other models.

**Comment by ParthSareen** [comment]
Hey everyone - sorry for the delay making my rounds right now. 

It is not due to relying on llama.cpp, I did have a branch working back early in Jan. As you may know we've been working on a new go engine. But since, all of us have been pretty heads down on that for the last month. There's going to be some bifurcation in running the new engine vs. current which means that new API features would need to have parity on both engines. In order to do that, just need to make sure the behavior between the old and new engines are the same for the logprobs endpoint as well as the tokenize/detokenize.

I've also been mainly working on our new sampling interfaces and the logprobs feature has been top of mind as well as I build it. I do hope to get to it soon I know it's super important to you all! The new engine is going to bring a lot of stability and maintainability throughout so we can support these kind of features much faster in the future.

**Comment by codelion** [comment]
> > It may take a while to get it here since they rely on llama.cpp underneath. You can try an alternative like optillm - https://github.com/codelion/optillm
> > 
> > * [Provide logits or logprobs in the API #2415 (comment)](https://github.com/ollama/ollama/issues/2415#issuecomment-2453442288)
> 
> Seems like this one is an openai tooling, while we're interested in other models.

This uses the same format and API as OpenAI but works for any model from hugging face.

**Comment by chaoyupeng** [comment]
Hi, Ollama team, any updates on the logprobs functionality updates?

**Comment by BruceMacD** [comment]
I've taken over implementation of this, expect some progress soon.

**Comment by BruceMacD** [comment]
Update for those interested:
I've opened some pull requests that refactor the model runners to make it possible to get information such as logprobs:
https://github.com/ollama/ollama/pull/9282

One that gets in I'll move forward with returning the values from the Ollama server.

**Comment by SabaPivot** [comment]
> Update for those interested: I've opened some pull requests that refactor the model runners to make it possible to get information such as logprobs: [#9282](https://github.com/ollama/ollama/pull/9282)
> 
> Once that gets in I'll move forward with returning the values from the Ollama server.

Cool! This is really cool!

**Comment by BruceMacD** [comment]
I'm still working on this at the same time as a few other things, but I haven't forgotten about it. Updated the #9282 pull request, so hopefully that one gets in soon.

**Comment by K0IN** [comment]
Hei, i know this might be the wrong thread, and it is surly somewhere documented, but why did ollama switch from the llama.cpp server to a custom implementation in the first place?

**Comment by aakash232** [comment]
Hello, Any updates on this feature?

**Comment by qiangzhou7** [comment]
Thank you very much for the efforts of the developers. Is there any update to this feature? This is a very important feature.

**Comment by BruceMacD** [comment]
Still working on it! I've take a diversion to work on some other stuff but more steps towards there are still in my short-term plans.

**Comment by brunodifranco** [comment]
Hello! Any estimate on when the feature will be completed?

**Comment by Proteusiq** [comment]
> Still working on it! I've take a diversion to work on some other stuff but more steps towards there are still in my short-term plans.

How can we help?

**Comment by CodeHatchling** [comment]
Please consider merging in pull request #9282 that implements this very needed feature.

**Comment by BarryKeee** [comment]
Any update on this? This feature is very much needed!

**Comment by enochlev** [comment]
May I create a statement of motivation. The completion of this feature is critical for a wide range of research. Enabling Ollama to expose raw logits allows researchers to extract high-quality supervision signals from large models like LLaMA 70B, even with limited hardware. This supports efficient knowledge distillation, making it possible to train smaller models with soft targets on modest GPUs like a single L40s or a 24GB gpus which I argue is the hardware limited to **95% of the researchers**. Unlocking this capability is a major contribution to the LLM research industry.

Most paid hosting LLMs prohibit the extraction of logits ([all except for OPENAI](https://docs.litellm.ai/docs/completion/input)), and we are limited to only using vLLM, which is quite GPU hungry.

Again, we understand this contribution comes from your free time, and it is much appreciated.

**Comment by SharmaM-dev** [comment]
[like]  Mridul Sharma reacted to your message:
________________________________

**Comment by baptistejamin** [comment]
I can be used for other things besides research. At Crisp, we utilize log-probs with fine-tuned models for binary classification on highly complex queries.

It can be super helpful.

```
Categorize this into two categories: SPAM, HAM

Query:
```

You can then have the log prob for SPAM, and use thresholds to classify.

**Comment by jaylinwylie** [comment]
Looking like we are waiting on the pull request to be approved. Unless theres a branch/fork we can experiment with in the meantime?

**Comment by unacceptable** [comment]
I am surprised that @jmorganca or @rick-github haven't closed this out yet saying to use a proxy.  That's what they did in #1053 and #8573 for auth and token usage.

**Comment by rick-github** [comment]
A proxy does not have access to logits.  See [here](https://huggingface.co/blog/logits-processor-zoo#what-are-logits-in-language-models) to learn more about logits.

**Comment by kjam** [comment]
Commenting to also say this is useful for implementing privacy and security controls, such as differential privacy predictions and regularization when managing adversarial input. Would be nice to merge https://github.com/ollama/ollama/pull/9282

**Comment by codelion** [comment]
> A proxy does not have access to logits. See [here](https://huggingface.co/blog/logits-processor-zoo#what-are-logits-in-language-models) to learn more about logits.

Depends on how it is implemented, OptiLLM is also a proxy but has an inbuilt local inference server and supports logits - https://github.com/codelion/optillm/issues/182

**Comment by Tritonio** [comment]
> > A proxy does not have access to logits. See [here](https://huggingface.co/blog/logits-processor-zoo#what-are-logits-in-language-models) to learn more about logits.
> 
> Depends on how it is implemented, OptiLLM is also a proxy but has an inbuilt local inference server and supports logits - [codelion/optillm#182](https://github.com/codelion/optillm/issues/182)

Is optiLLM able to get log-probs from Ollama though? I may be wrong but from a cursory look at the code it looks like it uses torch internally to get the logits, so if that is the case it's not acting as a proxy in front of ollama when it does so.

**Comment by codelion** [comment]
> Is optiLLM able to get log-probs from Ollama though?

No it is not through Ollama, you do not need ollama you can directly do the inference in OptiLLM with the in-built server which provides full logits via the standard OpenAI compatible API.

**Comment by rick-github** [comment]
The issue is about getting logits from ollama, not optillm.  Please don't distract from the issue.

**Comment by codelion** [comment]
> The issue is about getting logits from ollama, not optillm. Please don't distract from the issue.

OptiLLM is just an alternative, this issue has been open for over 18 months, if ollama wanted to implement it, they would have done it by now.

**Comment by rick-github** [comment]
Feel free to use optillm.  Others would like to get logits from ollama.  If you want to discuss alternatives, open a new issue.  This issue is for getting logits from ollama.

**Comment by SharmaM-dev** [comment]
[like]  Mridul Sharma reacted to your message:
________________________________

**Comment by VKrishna04** [comment]
yes please add it

**Comment by CodeHatchling** [comment]
I figured I'd throw in a couple of the many possible use cases for a feature like this.

Suppose you wanted the LLM to perform in a multiple choice type situation, or any other scenario where you want the model to determine the best-fitting response from a finite selection of options. The cleanest way to do this would be to evaluate the probability of each option and select the highest scoring one. This eliminates the need to handle unexpected responses from the model.

Another case is assisted writing, where the top N continuations are offered instead of the usual single continuation, such as with my tokenscape project on github.

Cheers!

**Comment by martinkozle** [comment]
> Suppose you wanted the LLM to perform in a multiple choice type situation, or any other scenario where you want the model to determine the best-fitting response from a finite selection of options. The cleanest way to do this would be to evaluate the probability of each option and select the highest scoring one. This eliminates the need to handle unexpected responses from the model.

Ollama does have structured outputs which you can use for this use-case.

https://github.com/ollama/ollama/blob/main/docs/openai.md#structured-outputs
https://ollama.com/blog/structured-outputs

In the Pydantic model use an enum with the valid options.

Not exactly what you said and doesn't excuse the lack of logits feature, but it may be useful if you need it for this specific case.

**Comment by Abdulrahman392011** [comment]
hey guys, so after this long read. I can't really tell when i can expect this to be available. 

I also want to say that ollama already provide the logprobs for openai-python compatibility layer. so I don't really understand what's the hold up. logically speaking it should be almost copy and paste with some modifications. can someone please correct me.

**Comment by rick-github** [comment]
Ollama does not make logprobs available, via the ollama API or the OpenAI compatible API.

**Comment by Abdulrahman392011** [comment]
> Ollama does not make logprobs available, via the ollama API or the OpenAI compatible API.

yes you're right i mistakenly saw the logprobs in the 
https://docs.ollama.com/openai
but it was unchecked. sorry for the confusion.

**Comment by Abdulrahman392011** [comment]
so what is needed to do this. why is it not implemented yet. I know that ollama runs a private instance of llama.cpp and llama.cpp does provide logprobs. so in my mind it's more like a command that just need modification when llama.cpp instance is initiated. 

I am here to learn, so correct me please.

**Comment by rick-github** [comment]
> I know that ollama runs a private instance of llama.cpp and llama.cpp does provide logprobs.

This issue was opened before llama.cpp provided OpenAI compatible logprobs.  In the meantime, ollama has migrated away from llama.cpp as the primary backend.  Work to support logprobs needs to be done on the new ollama engine.  The main developers are busy with other tasks.

**Comment by baptistejamin** [comment]
I just relesed this PR containing logprogs: https://github.com/ollama/ollama/pull/12899

You can try with:

```
GOTOOLCHAIN=auto go build .
```

```
./ollama serve
```

```
 curl http://localhost:11434/api/generate -d '{
                                                               "model": "llama3.2",
                                                               "prompt": "The capital of France is",
                                                               "stream": false,
                                                               "logprobs": true,
                                                               "top_logprobs": 1
                                                             }'
```

**Comment by tobiaswuerth** [comment]
+1

**Comment by rick-github** [comment]
https://github.com/ollama/ollama/releases/tag/v0.12.11

**Comment by jmorganca** [comment]
Wanted to say a huge thanks to @baptistejamin for the PR that got this in! And thank you to @BruceMacD who did some original work around this, @jessegross for the reviews and @ParthSareen for some fit and finish on the feature 🎉 

Thanks for closing this @rick-github 😊

**Comment by neuhaus** [comment]
I came across this new Ollama API feature while wondering how to implement the function described in the paper
"[LLMs can hide text in other text of the same length" (Norelli & Bronstein, 2024/2025)](https://arxiv.org/abs/2510.20075)" using the API.

The logprobs and top_logprobs endpoints return only the top N most likely tokens. I believe the API does not provide the full probability distribution over the entire vocabulary, nor does it allow efficient querying of a specific token's rank if it falls outside the top N.

I was wondering if this missing functionality could be added to the API.

**Comment by codelion** [comment]
> The logprobs and top_logprobs endpoints return only the top N most likely tokens. I believe the API does not provide the full probability distribution over the entire vocabulary, nor does it allow efficient querying of a specific token's rank if it falls outside the top N.

You can use OptiLLM for it if you want, it provides the full API - https://github.com/algorithmicsuperintelligence/optillm/issues/182

**Comment by rick-github** [comment]
To avoid confusion, OptiLLM is an alternative to ollama, it does not provide the full probability distribution for ollama models.

**Comment by Bottlecap202** [comment]
Integrate "memlayer" locally

**Comment by Bottlecap202** [comment]
And use a coral edge TPU, they are public now.

**Comment by baptistejamin** [comment]
Just to better understand. Would you like to build a product/inference system in production from this Paper?

IMO, what you need is something low-level API, such as LLama CPP, which is made for this.

Ollama CPP philosophy is to be similar to OpenAI API, and be easily plug & play

**Comment by Bottlecap202** [comment]
Kobold cpp.

**Comment by Bottlecap202** [comment]
Build around llama.cpp, BUT EASIER TO USE.

**Comment by Bottlecap202** [comment]
AND it has a community based horde. For all ages.

**Comment by Bottlecap202** [comment]
Kobold lite
